examples/sycl/pvc/pvc_gemm_mixed_dtype.cpp

```diff
@@ -0,0 +1,438 @@
+            Tile<_64,_64,_32>>; // Subgroup level-tile
+
+    constexpr int PipelineStages = 3;
+    using GEMMDispatchPolicy = cutlass::gemm::MainloopIntelPVC<PipelineStages>;
+    using EpilogueDispatchPolicy = cutlass::epilogue::IntelPVCEpilogue;
+
+    using EpilogueOp = cutlass::epilogue::fusion::LinearCombination<ElementOutput, ElementCompute,
+            ElementAccumulator, ElementAccumulator, cutlass::FloatRoundStyle::round_to_nearest>;
+
+    using FusionCallBacks = cutlass::epilogue::fusion::FusionCallbacks<EpilogueDispatchPolicy, EpilogueOp, TileShape,
+            decltype(tile_shape(TiledMma()))>;
+
+    using CollectiveEpilogueRef = cutlass::epilogue::collective::CollectiveEpilogue<
+            EpilogueDispatchPolicy,
+            TileShape,
+            ElementAccumulator,
+            cutlass::gemm::TagToStrideC_t<LayoutC>,
+            ElementOutput,
+            cutlass::gemm::TagToStrideC_t<LayoutD>,
+            FusionCallBacks,
+            XE_2D_U32x8x16_LD_N,
+            void, void,
+            XE_2D_U32x8x16_ST_N,
+            void, void>;
+
+    // Mainloop
+    using CollectiveMainloopRef = cutlass::gemm::collective::CollectiveMma<
+            GEMMDispatchPolicy,
+            TileShape,
+            ElementRefA,
+            cutlass::gemm::TagToStrideA_t<LayoutA>,
+            ElementB,
+            cutlass::gemm::TagToStrideB_t<LayoutB>,
+            TiledMma,
+            GmemTiledCopyA, void, void, cute::identity,  // A
+            GmemTiledCopyB, void, void, cute::identity   // B
+    >;
+
+    using GemmKernelRef = cutlass::gemm::kernel::GemmUniversal<
+    Shape<int, int, int, int>,
+    CollectiveMainloopRef,
+    CollectiveEpilogueRef
+    >;
+
+    using GemmRef = cutlass::gemm::device::GemmUniversalAdapter<GemmKernelRef>;
+
+    typename GemmRef::Arguments arguments{
+      cutlass::gemm::GemmUniversalMode::kGemm,
+      {options.m, options.n, options.k, options.l},
+      {block_A_dq.get(), stride_A, block_B.get(), stride_B},
+      {{options.alpha, options.beta}, block_C.get(), stride_C, block_ref_D.get(), stride_D}
+    };
+
+    // Run the gemm where the scaling is performed outside of the kernel.
+    GemmRef gemm_ref;
+    size_t workspace_size = GemmRef::get_workspace_size(arguments);
+    cutlass::device_memory::allocation<uint8_t> workspace(workspace_size);
+    CUTLASS_CHECK(gemm_ref.can_implement(arguments));
+    CUTLASS_CHECK(gemm_ref.initialize(arguments, workspace.get()));
+    CUTLASS_CHECK(gemm_ref.run());
+
+    // compare_reference
+    ElementOutput const epsilon(1e-2f);
+    ElementOutput const non_zero_floor(1e-4f);
+    bool passed = cutlass::reference::device::BlockCompareRelativelyEqual(block_ref_D.get(), block_D.get(), block_D.size(), epsilon, non_zero_floor);
+    return passed;
+  }
+
+  template <typename T1, typename T2> 
+  void initialize_block_A(cutlass::DeviceAllocation<T1>& block_device, cutlass::DeviceAllocation<T2>& block_device_dq,  uint64_t seed) {
```

Comment:
Do you know if this takes long? I've seen stuff like this slow down tests in the past (that was for GBs of data admittedly).
If you look [here](https://en.cppreference.com/w/cpp/numeric/random) you can see that mersenne twister algorithms are usually slower and I don't think we need "high quality" random numbers, whatever that means.
Depending on if it's an issue, it may be better to use oneMKL random libraries. You could produce the values on the device, copy them over and dequantise, then copy back.

Reply:
The example runs very quickly, so I suspect it's not an issue. I've changed to a 'subtract with carry' engine (`ranlux24_base `) because you're right that we don't really need quality random numbers. I agree it'd be more elegant to do this on the device, but given that we are trying to shift to validation via unit testing (and the test doesn't use this method) I'm reluctant to spend more effort on it.